docs/decisions/0004-error-handling.md

```diff
@@ -0,0 +1,59 @@
+
+- **Improper exception usage**. Some SK components use custom SK exceptions instead of standard .NET exceptions to indicate invalid arguments, configuration issues, and so on. This deviates from the standard approach for error handling in .NET and may frustrate SK client code developers.
+
+- **Exception hierarchy inconsistency**. Half of the custom SK exceptions are derived from SKException, while the other half are directly derived from Exception. This inconsistency in the exception hierarchy does not contribute to a cohesive exception model.
+
+- **Missing original exception details**. Certain SK exceptions do not preserve the original failure or exception details and do not expose them through their properties. This omission prevents SK client code from understanding the problem and handling it properly.
+
+- **No consolidated way for exception handling**. In some cases, SK has an exception type per component implementation, making it impossible for SK client code to handle them in a consolidated manner. Instead of having a single catch block, SK client code needs to include a catch block for each component implementation. Moreover, SK client code needs to be updated every time a new component implementation is added or removed.
+
+## Solution
+
+To address the identified issues, the following steps can be taken:
+
+- **Exception propagation**: To enable SK exceptions to propagate to SK client code, all SK components that currently do not do so should be modified to rethrow the exception instead of suppressing it. This allows SK client code to use try/catch blocks for handling and catching exceptions.
+
+- **Improper exception usage**: Modify SK code to throw .NET standard exceptions, such as ArgumentOutOfRangeException or ArgumentNullException, when class argument values are not provided or are invalid, instead of throwing custom SK exceptions. Analyze SK exception usage to identify other potential areas where standard .NET exceptions can be used instead.
+
+- **Exception hierarchy inconsistency**: Refactor SK custom exceptions that are not derived from SKException to derive from SKException. This ensures consistency in the SK exception hierarchy.
+
+- **Missing original exception details**: Identify all cases where the original exception or failure is not preserved either as an inner exception or as a property/message of the rethrown exception, and address them. For example, AI service connectors that use HttClient should preserve the original status code by either setting it as an inner exception or including it as part of the rethrown exception's property.
```

Comment:
> For example, AI service connectors that use HttClient should preserve the original status code by either setting it as an inner exception or including it as part of the rethrown exception's property.

Even if the original status code is copied out to whatever top-level exception gets thrown, there's still value in including the original exception as the inner exception.  In particular, any evolution of information exposed on that exception type will be available, for example if/when HttpRequestException is augmented with an HttpRequestError property as in https://github.com/dotnet/runtime/issues/76644#issue-1396942718.

Reply:
Yep, good point, updated the ADR.